### PR TITLE
fix# 273601: Updating Pianoroll Editor UI

### PR DESCRIPTION
**Pull request by blackears**
fix# 273601: Updating the Pianoroll Editor UI.

(This is a resubmission of https://github.com/musescore/MuseScore/pull/3781 which has been abandoned due to git issues.)

I've made some changes to the piano roll editor to update the UI to more closely resemble that of other music programs. I've replaced the transform() scaling which causes scaling artifacts with a system where only the coordinates are translated which allows for sharper drawing of shapes. I've also given each degree of the scale its own bar line. Also, it is now possible to click and drag on notes to change their pitch.

Many of the changes are preparation for allowing the user to edit note temporal aspects with the mouse too.

I've uploaded a quick demo:
https://youtu.be/-nsRiEoVfqA

Here's the old commit comments:
---

Updating main piano roll display to have 12 rows per octave. 

Updating piano keyboard to track main piano roll window.

Moving PianoKeyboard to its own file.

Better tracking of playhead during playback.

Drawing beat lines within measure.

Scrollbars now adjusted for stasis when zooming in the pianoview.

Piano ruler now showing tick marks respecting zoom level.

Fixing scroll issue during playback.

Taking direct control of selection.

PianoItem is no longer a Qt object.

Prepping  for note group drag.

Can now drag pitch changes.

Adjusting viewport pos when staff set.

Now displaying drumset names in PianoKeyboard when relevant.

### CODE REVIEW

**Comment by anatoly-os** [comment]
Hooray! Thank you. Btw, you can edit the description to avoid some not necessary messages like "removed cmakelists, added cmakelists"

**Comment by lasconic** [comment]
I just compiled the PR to try it.
I get the following messages in the log
```
unknown:unknown: QObject::connect: No such signal Ms::PianoView::posChanged(const Pos&) in mscore/pianoroll.cpp:186
unknown:unknown: QObject::connect: No such signal Ms::PianoView::posChanged(const Pos&) in mscore/pianoroll.cpp:187
```

----

Also, I got a crash in `PianoItem::paint()`
on `painter->drawText(textHiliteRect, Qt::AlignLeft | Qt::AlignTop, pitchNames[pitch % 12]);`

1. Using untitled score, add 4 quarters C,D,E,F
2. Right click > Piano roll editor
3. Press space > Crash

----

In pianoview.cpp the indent is a mess. Have a look to our [code style guide](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules) and check how other files are indented. In particular, the braces.

File: mscore/musescore.cpp [hunk 1]
```diff
@@ -3893,13 +3893,18 @@ void MuseScore::handleMessage(const QString& message)
...
       {
       if (pianorollEditor == 0)
             pianorollEditor = new PianorollEditor;
       pianorollEditor->setScore(staff->score());
       pianorollEditor->setStaff(staff);
       pianorollEditor->show();
+      
+      //vbox 98
+//            printf("editInPianoroll type:%d\n", focus == 0 ? -1 : (int)focus->type());
+
```
Comment (Jojo-Schmitz) [inline mscore/musescore.cpp]: better lose those debugging artifacts

File: mscore/pianoroll.cpp [hunk 2]
```diff
@@ -498,7 +498,9 @@ void PianorollEditor::heartBeat(Seq* seq)
       if (locator[0].tick() != tick) {
             posChanged(POS::CURRENT, tick);
             if (preferences.getBool(PREF_APP_PLAYBACK_FOLLOWSONG))
+                  {
                   gv->ensureVisible(tick);
+                  }
```
Comment (Jojo-Schmitz) [inline mscore/pianoroll.cpp]: No need for those braces. But if, the opening one should be a line up

File: mscore/scoreview.cpp [hunk 3]
```diff
@@ -4181,6 +4184,42 @@ static bool elementLower(const Element* e1, const Element* e2)
...
+      int bestTick = 0;
+      Element* bestEle = 0;
+//      printf("bestTick page:%d\n", page->no());
+      
+      double w = (preferences.getInt(PREF_UI_CANVAS_MISC_SELECTIONPROXIMITY) * .5) / matrix().m11();
+      QRectF r(p.x() - w, p.y() - w, 3.0 * w, 3.0 * w);
+
+      QList<Element*> el = page->items(r);
+      for (Element* e : el) {
+//            printf("examining type:%d tick:%d\n",  (int)e->type(), (int)e->tick());
```
Comment (Jojo-Schmitz) [inline mscore/scoreview.cpp]: here too better lose those debugging artifacts

File: mscore/scoreview.h [hunk 4]
```diff
@@ -153,7 +153,8 @@ class ScoreView : public QWidget, public MuseScoreView {
       void paint(const QRect&, QPainter&);
 
       void objectPopup(const QPoint&, Element*);
-      void measurePopup(const QPoint&, Measure*);
+//      void measurePopup(const QPoint&, Measure*);
```
Comment (Jojo-Schmitz) [inline mscore/scoreview.h]: better delete, git history is enough to show it was once there

Comment (blackears) [reply mscore/scoreview.h]: This isn't actually my change.  Not sure how it got here.

File: mscore/musescore.cpp [hunk 5]
```diff
@@ -3900,10 +3900,6 @@ void MuseScore::editInPianoroll(Staff* staff, Element* focus)
       pianorollEditor->setScore(staff->score());
       pianorollEditor->setStaff(staff);
       pianorollEditor->show();
-      
-      //vbox 98
-//            printf("editInPianoroll type:%d\n", focus == 0 ? -1 : (int)focus->type());
-
```
Comment (Jojo-Schmitz) [inline mscore/musescore.cpp]: better lose those debugging artifacts

File: CMakeLists.txt [hunk 6]
```diff
@@ -710,7 +710,7 @@ add_custom_target(lrelease
...
 ##
 
-add_custom_command(OUTPUT ${PROJECT_BINARY_DIR}/plugins
-   COMMAND genManual ${PROJECT_SOURCE_DIR} ${PROJECT_BINARY_DIR}
-   DEPENDS genManual
-   )
+#add_custom_command(OUTPUT ${PROJECT_BINARY_DIR}/plugins
+#   COMMAND genManual ${PROJECT_SOURCE_DIR} ${PROJECT_BINARY_DIR}
+#   DEPENDS genManual
+#   )
```
Comment (Jojo-Schmitz) [inline CMakeLists.txt]: This change needs to get reverted

File: mscore/pianoroll.cpp [hunk 7]
```diff
@@ -228,12 +226,30 @@ PianorollEditor::~PianorollEditor()
...
 //---------------------------------------------------------
 //   setStaff
 //---------------------------------------------------------
 
 void PianorollEditor::setStaff(Staff* st)
       {
+      if (staff == st)
+            return;
+
+
```
Comment (Jojo-Schmitz) [inline mscore/pianoroll.cpp]: too many empty lines

File: mscore/scoreview.cpp [hunk 8]
```diff
@@ -4181,6 +4184,40 @@ static bool elementLower(const Element* e1, const Element* e2)
...
+Element* ScoreView::nearestTickElement(QPointF p)
+      {
+      //Look through the score to find an element near the mouse click pos
+      // that can be used to determine the tick closest to the mouse position
+      
+      Page* page = point2page(p);
+      if (!page)
+            return 0;
+
+      p       -= page->pos();
```
Comment (Jojo-Schmitz) [inline mscore/scoreview.cpp]: too many spaces

File: mscore/pianoview.cpp [hunk 9]
```diff
@@ -291,155 +360,334 @@ void PianoView::createLocators()
...
-            // if xpos <= 0, then the scene is centered
-            // there is no scroll bar anymore sending
-            // change signals, so we have to do it here:
-            //
-            double xpos = -(mapFromScene(QPointF()).x());
-            if (xpos <= 0)
-                  emit xposChanged(xpos);
+
+      if (event->modifiers() == Qt::ControlModifier) 
+            {
```
Comment (Jojo-Schmitz) [inline mscore/pianoview.cpp]: brace should go to line above

File: mscore/pianoview.cpp [hunk 10]
```diff
@@ -291,155 +360,334 @@ void PianoView::createLocators()
...
+            updateNotes();
+            
+            int mousePixX = tickToPixelX(mouseXTick);
+            horizontalScrollBar()->setValue(mousePixX - event->x());
+
+            scene()->update();
             }
-      else if (event->modifiers() == Qt::ShiftModifier) {
+      else if (event->modifiers() == Qt::ShiftModifier) 
+            {
```
Comment (Jojo-Schmitz) [inline mscore/pianoview.cpp]: ditto

File: mscore/pianoview.cpp [hunk 11]
```diff
@@ -291,155 +360,334 @@ void PianoView::createLocators()
...
+            //Horizontal scroll
             QWheelEvent we(event->pos(), event->delta(), event->buttons(), 0, Qt::Horizontal);
             QGraphicsView::wheelEvent(&we);
             }
-      else if (event->modifiers() == 0) {
+      else if (event->modifiers() == 0)
+            {
```
Comment (Jojo-Schmitz) [inline mscore/pianoview.cpp]: ditto, but not needed at all here

**Comment by blackears** [comment]
While I do not get a crash, I have noticed that when playback is engaged that `_note->pitch() + _event->pitch()` is often not equal to the pitch value.  While I think I can fix the UI by simply caching the pitch value then when the note bounds are calculated, this does suggest a deeper bug within Musescore's note calculation mechanism.

**Comment by blackears** [comment]
In the older version of this editor, each visible note block in the UI coresponded to a (Note, NoteEvent) tuple and was calculated and cached whenever the updateNotes() method was called.  However, the cached NoteEvents would sometimes be deallocated (this sometimes happened when the user made a new selection) and attempting to access them afterward would result in strange behavior.  As a result, I switched the display to represent the Notes directly. This affects how some notes are displayed (their score length, rather than their playback length).  Also, when a user need to update the playback data stored on the NoteEvent, an up-to-date copy of the NoteEvent is fetched rather than relying on the cached copy.

**Comment by MarcSabatella** [comment]
Hmm.  I can't imagine not showing actual note events would be a good thing - that seems to defeat much of the purpose of the editor.  Actual on times, actual durations, and the actual multiple events associated with ornaments etc.  Otherwise, what is the piano roll editor but a different UI for doing the same sort of things that could be done more directly in notation?  Or am I misunderstanding something?

**Comment by blackears** [comment]
Multiple views of the same data can be a very useful thing!  Different tools are more efficient at different tasks, and being able to switch between them lets you get to your objective faster.

That said, I could add a button that could switch between a score view and a play event view.  The old editor wasn't handling the play view properly anyway due to NoteEvents being deallocated and the NoteUp/NoteDown actions operating on the Notes anyway.

**Comment by MarcSabatella** [comment]
I guess the big question here to me is, what is the actual point this PR?  If the goal is to make it easier to do things that actually *require* use of the piano roll editor (editing playback details), then I am all for it.  If it's to make the piano roll into another way to edit the notation, I'm unconvinced.

**Comment by blackears** [comment]
I've reworked things so that the editor is now showing playback data again and the functionality is the same as it was in the old editor.

**Pull request closed without merging**